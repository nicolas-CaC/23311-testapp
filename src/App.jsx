import { useState } from 'react'
import './App.css'
import { useFetch } from './hooks/useFetch'

const initialValue = 0


// eslint-disable-next-line react/prop-types
export const App = ({ funcion }) => {

    const [value, setValue] = useState(initialValue)
    const { loading } = useFetch()

    const sumar = () => setValue(value + 1)
    const restar = () => setValue(value - 1)
    const reset = () => setValue(initialValue)

    const handlechange = ({ target }) => setValue(target.value)

    const onSubmit = (e) => {
        e.preventDefault()
        if (value.trim() <= 1) return
        setValue('')
        funcion(value.trim())
    }

    return (
        <div>
            <p data-testid='state'>{ value }</p>
            <form onSubmit={ onSubmit } aria-label='form'>
                <input
                    type="text"
                    onChange={ handlechange }
                    placeholder='Valor'
                    value={ value }
                />
            </form>

            <button data-testid='resta' onClick={ restar }>Restar</button>
            <button aria-label='reset' onClick={ reset }>Reset</button>
            <button data-testid='suma' onClick={ sumar }>Sumar</button>

            <hr />
            <img src="https://picsum.photos/200/300" alt="imagen" />
            <div>{ loading ? 'Cargando datos...' : 'Cargado' }</div>

        </div>
    )
}

export default App
