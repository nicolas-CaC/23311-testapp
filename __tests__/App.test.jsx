import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, test, vitest } from "vitest";
import App from "../src/App";


describe('Test: App.jsx', () => {

    // const title = 'Titulo'

    test('toMatchSnapshot', () => {
        const { container } = render(<App />)
        expect(container).toMatchSnapshot()
    })

    test('getByText', () => {
        const { getByText } = render(<App />)
        expect(getByText('Restar')).toBeTruthy()
    })

    test('toContain', () => {
        const { container } = render(<App />)
        const divs = container.getElementsByTagName('div')
        expect(divs[0].innerHTML).toContain('Cargando datos')
    })

    test('getByTestId', () => {
        const { getByTestId } = render(<App />)
        expect(getByTestId('state')).toBeTruthy
    })

    test('getAllByText', () => {
        const { getAllByText } = render(<App />)
        expect(getAllByText('Sumar')).toBeTruthy()
    })

    test('getAllByTestId', () => {
        const { getAllByTestId } = render(<App />)
        expect(getAllByTestId('state').length).toBeGreaterThanOrEqual(1)
    })

    test('screen', () => {
        render(<App />)
        expect(screen.getAllByTestId('resta').length).toBeGreaterThanOrEqual(1)
    })

    test('valor inicial 0', () => {
        render(<App />)
        expect(screen.getByText(0)).toBeTruthy()
    })

    test('Eventos: Boton Suma', () => {
        render(<App />)
        fireEvent.click(screen.getByTestId('suma'))
        expect(screen.getByText(1)).toBeTruthy()
    })

    test('Eventos: Boton Resta', () => {
        render(<App />)
        fireEvent.click(screen.getByTestId('resta'))
        expect(screen.getByText(-1)).toBeTruthy()
    })

    test('Eventos: Boton Reset', () => {
        render(<App />)
        fireEvent.click(screen.getByTestId('resta'))
        // screen.debug()
        fireEvent.click(screen.getByRole('button', { name: 'reset' }))
        // screen.debug()
        expect(screen.getByTestId('state').innerHTML).toBe('0')
    })

    test('Desestructuracion', () => {

        render(<App />)

        const { src, alt } = screen.getByRole('img')

        expect(src).toMatch('https://picsum.photos/200/300')
        expect(alt).toBe('imagen')
    })

    test('Form & Input', () => {

        const funcionMock = vitest.fn()

        render(<App funcion={ funcionMock } />)

        const form = screen.getByRole('form')
        const input = screen.getByRole('textbox')

        fireEvent.input(input, { target: { value: 4 } })
        expect(input.value).toBe('4')

        fireEvent.submit(form)
        expect(input.value).toBe('')

        expect(funcionMock).toHaveBeenCalled()
        expect(funcionMock).toHaveBeenCalledTimes(1)
        expect(funcionMock).toHaveBeenCalledWith('4')
    })






})