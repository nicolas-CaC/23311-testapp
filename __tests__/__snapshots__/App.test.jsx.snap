// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: App.jsx > toMatchSnapshot 1`] = `
<div>
  <div>
    <p
      data-testid="state"
    >
      0
    </p>
    <form
      aria-label="form"
    >
      <input
        placeholder="Valor"
        type="text"
        value="0"
      />
    </form>
    <button
      data-testid="resta"
    >
      Restar
    </button>
    <button
      aria-label="reset"
    >
      Reset
    </button>
    <button
      data-testid="suma"
    >
      Sumar
    </button>
    <hr />
    <img
      alt="imagen"
      src="https://picsum.photos/200/300"
    />
    <div>
      Cargando datos...
    </div>
  </div>
</div>
`;
